cips/cip-shwap_protocol.md

```diff
@@ -0,0 +1,407 @@
+This ongoing exchange ensures that as soon as any peer acquires the sought-after data, it can instantly share it with
+those in need.
+
+Shwap is designed to be synergetic with Bitswap, as that is the primary composition to be deployed in Celestia's DA
+network. Bitswap provides the 1/N peers guarantee and can parallelize fetching across multiple peers. Both of these properties
+significantly contribute to Celestia's efficient DAS protocol.
+
+Bitswap runs over the libp2p stack, which provides QUIC transport integration. Subsequently, Shwap will benefit from features
+libp2p provides together with transport protocol advancements introduced in QUIC.
+
+#### Multihashes and CID
+
+Bitswap is tightly coupled with [Multihash][mh] and [CID][cid] notions, establishing the [content addressability property][content-address].
+Bitswap operates over Blocks of data that are addressed and verified by CIDs. Based on that, Shwap integrates into
+Bitswap by complying with both of these interfaces. The [Share Containers](#share-containers) are Blocks that are identified
+via [Share Identifiers](#share-identifiers).
+
+Even though Shwap takes inspiration from content addressability, it breaks free from the hash-based model to optimize
+message sizes and data request patterns. In some way, it hacks into multihash abstraction to make it contain data that
+is not, in fact, a hash. Furthermore, the protocol does not include hash digests in the multihashes. The authentication
+of the messages happens using externally provided data commitment.
+
+However, breaking-free from hashes creates issues necessary to be solved on the implementation level, particularly in 
+[the reference Golang implementation][gimpl], if forking and substantially diverging from the upstream is not an option. 
+CIDs are required to have fixed and deterministic sizes. Making share identifiers compliant with CID 
+prevents protobuf usage due to its reliance on varints and dynamic byte arrays serialization in.
+
+The naive question would be: "Why not make content verification after Bitswap provided it back over its API?" Intuitively,
+this would simplify much and would not require "hacking" CID. However, this has an important downside - the Bitswap, in 
+such a case, would consider the request finalized and the content as fetched and valid, sending a DONT_WANT message to 
+its peers. In contrast, the message might still be invalid according to the verification rules.
+
+Bitswap still requires multihashes and CID codecs to be registered. Therefore, we provide a table for the
+supported [share identifiers](#share-identifiers) with their respective multihash and CID codec codes. This table
+should be extended whenever any new share identifier is added.
+
+| Name     | Multihash | Codec  |
+|----------|-----------|--------|
+| RowID    | 0x7801    | 0x7800 |
+| SampleID | 0x7811    | 0x7810 |
+| DataID   | 0x7821    | 0x7820 |
+
```

Comment:
Does this also imply, that incompatible changes to CID messages would also require introduction new multihash code and codec code

Reply:
Yes, it's worth documenting that.

Reply:
Done